### DISCUSSION TITLE
Programatically setValue for single form value without subscribing to whole form?

### DISCUSSION THREAD

**Discussion by tetchel** [react-hook-form/react-hook-form#7752] [Q&A] [answered]
Hi, so I have been using the `useWatch` API in order to watch the minimal set of values when rendering a component.

It's my understanding that this way I do not have to `useFormContext` and thus have my component re-render every time any form value changes - i subscribe only to the fields I need, which is great.

However, sometimes I need to watch one or two values in a component and imperatively set one of those values if the other changes. Let me try and demonstrate with an example:

```typescript
const option1 = useWatch({ name: 'option1', defaultValue: false })
const option2 = useWatch({ name: 'option2', defaultValue: false }}

useEffect(() => {
  
  if (option1 === true) {
    // let's say I want to set option2 to true in this case.
  }
}, [ option1 ])

return (
  // render some form which allows setting option 1 and 2
)
```

Is there some way to access `form.setValue('option1')` without having to pull in the whole `form` context?
I am thinking of an ideal usage like a `useSetValue(<name>)` API which works similar to useWatch but returns a function where I can set a form value programatically.

Thanks for reading

**Comment by bluebill1049** [comment, marked as answer]
> However, sometimes I need to watch one or two values in a component and imperatively set one of those values if the other changes. Let me try and demonstrate with an example:

```tsx
useWatch({
  name: ['option1', 'option2']
})
```

> Is there some way to access form.setValue('option1') without having to pull in the whole form context?

not at this stage

**Comment by tetchel** [comment]
OK, thanks for the quick response. I hope to see that feature considered for future releases.

Does this note in `useFormState` proxy apply only to that hook? IE, not to `useFormContext` or `useForm`?
![image](https://user-images.githubusercontent.com/8888438/152631397-2d4aa2fb-1621-4969-8dfd-262511d6ec59.png)

**Reply by bluebill1049** [reply]
`useFormState` is at its own hook level. (not `useForm`)

**Reply by tetchel** [reply]
sorry I'm not sure I understand the answer. 

I'm asking if there is a performance benefit to deconstructing the `useFormContext` or `useForm` return, as there is a benefit for `useFormState`.

**Reply by bluebill1049** [reply]
yes, what are you deconstructing or reading is what's hook form going to subscribe to, and include the following: 

- re-render (when to update the state)
- computation (when to skip computation if irrelevant state is not subscribed)

**Reply by tetchel** [reply]
OK. that's really good to know. Maybe the API doc pages for those hooks should have a note describing that? As there is in the screenshot above for `useFormState`

**Reply by bluebill1049** [reply]
would you like to send a PR to improve on it?

**Reply by tetchel** [reply]
I can put up a PR. I am looking at [this paragraph](https://github.com/react-hook-form/documentation/blob/a0c700d341b76922b9f014d39a578f7043a88f66/src/components/UseFormState.tsx#L47).

Does it apply to all the hooks, except `useWatch`? 

ie:
[useForm](https://react-hook-form.com/api/useform)
[useController](https://react-hook-form.com/api/usecontroller)
[useFormContext](https://react-hook-form.com/api/useformcontext)
[useFormState](https://react-hook-form.com/api/useformstate) (already in this one)
[useFieldArray](https://react-hook-form.com/api/usefieldarray)

**Reply by bluebill1049** [reply]
only applicable to `formState`

so the following only

```
const {formState} = useForm();
const formState = useFormState();
```

**Reply by tetchel** [reply]
I'm confused. Then the documentation doesn't need an update?

**Reply by tetchel** [reply]
_oh_, I understand now. formState within those other hooks still benefits from the proxy, but the other returned values don't.

**Reply by bluebill1049** [reply]
That's right.

**Reply by bluebill1049** [reply]
I wonder which part is confusing which we can improve on it.

**Reply by tetchel** [reply]
I just misunderstood your comment above - https://github.com/react-hook-form/react-hook-form/discussions/7752#discussioncomment-2121687

I interpreted that eg. 

`const { setValue } = useFormContext()` 
may be more performant than 
`const form = useFormContext(); form.setValue()`

but I understand now that's not correct and you were  only referring to the `formState`.

**Reply by tetchel** [reply]
anyway thanks for the quick responses. Your work on this library is admirable :)

**Reply by bluebill1049** [reply]
most welcome.
